### ISSUE TITLE
Version 4

### ISSUE THREAD

**Issue by willfarrell** [middyjs/middy#769]
PR: #908 

## Planned
- [x] Drop Node v14, Add support fro Node v18
  - [x] Update `engines` to `"node": ">=16"`
  - [x] Update github workflows `node-version`
  - [x] Update `swc` to use `es2021` (https://github.com/tsconfig/bases/tree/main/bases)
- [x] `event-normalizer`: review for any new events from [AWS docs](https://docs.aws.amazon.com/lambda/latest/dg/lambda-services.html) and [event list](https://github.com/michaelbrewer/aws-lambda-events)
- [x] `http-security-headers`: review against helmet, MDN, ASVS
- [x] AWS SDK v3 #584
  - [x] Update packages
  - [x] Update tests
  - [x] Update docs
  - [x] TypeScript
  - [x] Update to use async iterators for pagination in `ssm`, `secrets-manager`, etc (https://aws.amazon.com/blogs/developer/pagination-using-async-iterators-in-modular-aws-sdk-for-javascript/)
- [x] `http-urlencode-body-parser` throw 422 when unable to parse
- [x] `http-content-encoding` pull stream piping into new middleware `stream-transform-response`
- [x] `http-router` parse path params. no, can be done in v3
- [x] `http-router` REST and HTTP route paths should be the same (https://github.com/middyjs/middy/issues/873). No, see issue.
- [x] `validator` export `compileSchema` and require it be passed in to allow for better tree-shaking
- [x] Allow top level await within middy core when adding middleware `.use(await middleware())`
- [x] Update bundling docs
 
### >= 15
- [x] use replaceAll in `event-normalizer` [skip - slower than current]

#### remove polyfills
- [x] `core`
  - [x] Remove `AbortControler` polyfill
  - [x] remove `timers/promises` polyfill
- [x] `http-content-encoding`: remove `stream/promises` polyfill
- [x] `util` update tests to use `timers/promises` for delay

#### Error {cause} >= 16.9
- [x] `util` update `createError` to use `class HttpError extends Error`
- [x] `http-json-body-parser`: use proper `new Error(message, {cause})` notation
- [x] `http-multipart-body-parser`: use proper `new Error(message, {cause})` notation
- [x] `util`: use proper `new Error(message, {cause})` notation
- [x] `util`: use proper `super(message, {cause})` notation in `HttpError`
- [x] `validator`: use proper `new Error(message, {cause})` notation

## deprecate
- [x] deprecate `cloudwatch-metrics` for `@aws-lambda-powertools/metrics` (uses EMF)? @fredericbarthelet
- [x] `validator`: deprecate `inputSchema` and `outputSchema` options
- [x] `event.rawBody` in `http-json-body-parser`

**Comment by vicary** [comment]
@willfarrell Does `.handler()` means it inherits all the event types from the `.use()` before it and I no longer have to chain up middleware events in my resolver?

My team is currently repeating the following style in hundreds of resolvers. Every night, I dream of making all the middleware events private and self-contained.

```typescript
const resolver = async (event: FooEvent & BarEvent & BazEvent) => { ... };
export const handler = middy(resolver).use(...).use(...).use(...);
```

**Comment by willfarrell** [comment]
@vicary this will be in v3. What happens is whatever is passed into `.handler()` replace the first arguments to `middy()` (likely nothing), then end the chaining.

It means you can do `export const handler = middy().use(...).use(...).handler(resolver)` instead.

Which mean it's easier to split up repeated middlewares:
```
// middlewares.js
export default middy().use(...).use(...)

// lambda.js
import wrapper from 'middleware.js'

const resolver = async (event: FooEvent & BarEvent & BazEvent) => { ... };
export const handler = wrapper.use(validator({...})).handler(resolver)
```

Technically you can do this in v2.x but you have to wrap it in a function, which isn't as easy to use.

**Comment by vicary** [comment]
@willfarrell Not sure if I get it.

I specifically wants to remove the necessity to chain `FooEvent & BarEvent & BazEvent` from my `resolver()` because they are events in the middlewares which add something to the end result. Can I do the following in v3?

```typescript
import type { AppSyncResolverEvent } from "aws-lambda";

export const handler = middy()
   .use(foo()) // does TEvent | FooEvent inside
   .use(bar()) // does TEvent | BarEvent inside
   .use(baz()) // does TEvent | BazEvent inside
   .handler<AppSyncResolverEvent>((event) => {
      // Can I automatically get all the followings?
      event.foo; // FooEvent
      event.bar; // BarEvent
      event.baz; // BazEvent
      event.arguments; // AppSyncResolverEvent
   });
```

**Comment by willfarrell** [comment]
middy is just a light framework to order repeated operation before/after a handler execution. What is in those middlewares is completely up to you to fit your needs.

All of the following are the same:
```
export const handler = middy().use(foo()).use(bar()).use(baz()).handler(...)
export const handler = middy().use([foo(), bar(), baz()]).handler(...)
export const handler = middy().use(foobarbaz()).handler(...)

import middlewares from '...'
export const handler = middlewares.handler(...)
```
Hope that help in your planning on how to best use middy.

Your above example looks like it will run just fine in v3. If you move the handler into `middy(...)` it can run in v2.

**Comment by vicary** [comment]
@willfarrell Main problem is TypeScript implementation does not carry/merge event types down the chain, it should be possible in the DT for `.use()` without too much effort.

My second code won't pass type checking because the handler type has no knowledge about `FooEvent`, `BarEvent` and `BazEvent`, it only knows the `AppSyncResolverEvent` I specifically give it.

I don't want to manually import all the intermediate middleware types in my handler `(event: A & B & C & D & E ........) => {}`, I want them to automatically be carried down when I call `.use()`.

**Comment by willfarrell** [comment]
Ooohhhh, this is a typescript questions. Sorry. I personally don't use typescript so I can't advise how to best use middy with typescript. I know you're not the first to run into this issue, I'm unsure how other have resolved this. I'm open to feedback on how to this can be improved within middy, if even possible.

**Comment by vicary** [comment]
It's mildly troubling, but somehow breaking how types work. Not sure if I have the time and energy to whip up a PR for the already launching v3, so I am speaking in this v4 thread.

**Comment by vicary** [comment]
That's great! I'll see what I can do next month.

**Comment by dbartholomae** [comment]
@vicary You might want to take a look at https://github.com/dbartholomae/lambda-middleware which I created out of a discussion around how to better use TypeScript types in middy. It attacks the problem by using functional composition, so middlewares can define their input and output types.

**Comment by oyed** [comment]
@dbartholomae Looks great - I recently recreated the middleware pattern for TypeScript too (https://github.com/tnotifier/lamware), but went a different route with chaining `.use()` to have pass-through typing for a custom state that the handler gets given

**Comment by vicary** [comment]
@willfarrell I can start from what I know, if you're OK with the approach in https://github.com/middyjs/middy/issues/589#issuecomment-1195096567.

In the same PR I'll also fix inline middlewares, because they are broken in a way that their `TEvent` are always `any`.

**Comment by willfarrell** [comment]
I look forward to reviewing the PR.

**Comment by nponeccop** [comment]
> deprecate use of qs in http-urlencode-body-parser for node:querystring

Shouldn't it be `new URLSearchParams`? It is both W3C-standardized and supported by node.

**Comment by codyfrisch** [comment]
> Ooohhhh, this is a typescript questions. Sorry. I personally don't use typescript so I can't advise how to best use middy with typescript. I know you're not the first to run into this issue, I'm unsure how other have resolved this. I'm open to feedback on how to this can be improved within middy, if even possible.

Yeah, TypeScript is important for me, and the fact its poorly implemented here is why I have chucked the last two days work trying to refactor to use Middy and am searching for a more competent solution.

**Comment by willfarrell** [comment]
@codyfrisch Sorry to hear Middy TypeScript definitions are not meeting your needs. A PR to help improve them is welcome if you'd like to contribute back to the community. I've talked with other OSS maintainers about supporting TypeScript well, nearly all are also struggling with it. This article was shared with me that I hope will shed light on this topic: https://erock.prose.sh/typescript-terrible-for-library-developers

**Comment by bilalq** [comment]
@willfarrell FWIW, the #883 improved the situation a ton. Really appreciate the work you and @vicary did to get that merged in! Honestly, at this point, I have no reservations about using Middy with TS. It's been pretty stellar.


Regarding your point about difficulty in supporting TypeScript as a library author: I agree that it is annoying if your source code itself isn't authored in TypeScript. However, I've found maintaining libraries and frameworks a lot easier when they're actually written in TS compared to what it was like back in the pure JS days. The article you linked to is definitely right about how the heavy need for complex type expressions in libraries vs end apps basically forces you to build up expertise on the type system if you want to have a smooth developer experience.

**Comment by vicary** [comment]
@bilalq To be fair, it's impractical to overhaul a project into if the authors are not already using TypeScript everyday.

But refactoring is happening by any chance, count me in for the help and long term maintenance.

**Comment by naorpeled** [comment]
Hey @willfarrell,
would love to help with some tasks,
I can start with `http-router - parse path params`,
if you have another task in mind let me know :)

**Comment by willfarrell** [comment]
@naorpeled Thanks for the offer! I've been rethinking whether to include this in v4 or just add to v3. Depending on how it's implemented I think it could be included into v3 as a minor update and not have to wait for v4 to release (date unknown - waiting on AWS). I was imagining it would parse by default with an option to disable for APIG use. Please feel free to open a PR for this off of the `main` branch.

**Comment by willfarrell** [comment]
From a feature standpoint, #908, is ready for an alpha release. Because NodeJS 18 runtime has not released yet, additional change may be applied later. Please take a look and give feedback.

If someone could review and update the TypeScript to accommodate the changes to support AWS SDK v3, that would be greatly appreciated.

Thanks

**Comment by naorpeled** [comment]
> From a feature standpoint, #908, is ready for an alpha release. Because NodeJS 18 runtime has not released yet, additional change may be applied later. Please take a look and give feedback.
> 
> If someone could review and update the TypeScript to accommodate the changes to support AWS SDK v3, that would be greatly appreciated.
> 
> Thanks

Hey,
I'd love to take this.

Regarding the task we've discussed earlier here, I did some digging into the event payloads in the different AWS services but I feel I need to look further into it, was on a spontaneous vacation so didn't have time to look further into it, I think I'll be able to work on it during this weekend.

**Comment by willfarrell** [comment]
@naorpeled I already knocked out the feature for parsing pathParams. I woke up with an idea on how to do it cleanly, so I just did it. It just needs to be released, something for tomorrow.

**Comment by naorpeled** [comment]
> @naorpeled No worries, we all need vacation. I already knocked out the feature for parsing pathParams. I woke up with an idea on how to do it cleanly, so I just did it. It just needs to be released, something for tomorrow.

👑

**Comment by willfarrell** [comment]
an initial alpha is not released for testing

**Comment by willfarrell** [comment]
nodejs18.x runtime is finally here. v4 will release soon after some additional testing.
